File: src/main/java/studio/blacktech/furryblackplus/core/common/time/TimeTool.java
/*
 * Copyright (C) 2021 Alceatraz @ BlackTechStudio
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the BTS Anti-Commercial & GNU Affero General
 * Public License as published by the Free Software Foundation, either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * BTS Anti-Commercial & GNU Affero General Public License for more details.
 *
 * You should have received a copy of the BTS Anti-Commercial & GNU Affero
 * General Public License along with this program.
 *
 */


package studio.blacktech.furryblackplus.core.common.time;

import studio.blacktech.furryblackplus.FurryBlack;
import studio.blacktech.furryblackplus.common.Api;

import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;


@Api("时间工具")
public class TimeTool {


    private static final DateTimeFormatter DATETIME = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss").withZone(ZoneId.systemDefault());
    private static final DateTimeFormatter DATE = DateTimeFormatter.ofPattern("yyyy-MM-dd").withZone(ZoneId.systemDefault());
    private static final DateTimeFormatter TIME = DateTimeFormatter.ofPattern("HH:mm:ss").withZone(ZoneId.systemDefault());


    private TimeTool() {}


    // ==================================================================================================


    @Api("格式化日期")
    public static String date() {
        return DATE.format(Instant.now());
    }

    @Api("格式化日期")
    public static String date(long timeStamp) {
        return DATE.format(Instant.ofEpochMilli(timeStamp));
    }

    @Api("格式化日期")
    public static String date(Instant instant) {
        return DATE.format(instant);
    }

    @Api("格式化日期")
    public static String date(LocalDateTime localDateTime) {
        return DATE.format(localDateTime);
    }


    @Api("格式化时间")
    public static String time() {
        return TIME.format(Instant.now());
    }

    @Api("格式化时间")
    public static String time(long timeStamp) {
        return TIME.format(Instant.ofEpochMilli(timeStamp));
    }

    @Api("格式化时间")
    public static String time(Instant instant) {
        return TIME.format(instant);
    }

    @Api("格式化时间")
    public static String time(LocalDateTime localDateTime) {
        return TIME.format(localDateTime);
    }


    @Api("格式化日期时间")
    public static String datetime() {
        return DATETIME.format(Instant.now());
    }

    @Api("格式化日期时间")
    public static String datetime(long timeStamp) {
        return DATETIME.format(Instant.ofEpochMilli(timeStamp));
    }

    @Api("格式化日期时间")
    public static String datetime(Instant instant) {
        return DATETIME.format(instant);
    }

    @Api("格式化日期时间")
    public static String datetime(LocalDateTime localDateTime) {
        return DATETIME.format(localDateTime);
    }


    @Api("自定义格式化")
    public static String format(String pattern) {
        return DateTimeFormatter.ofPattern(pattern).withZone(ZoneId.systemDefault()).format(Instant.now());
    }

    @Api("自定义格式化")
    public static String format(String pattern, ZoneId zone) {
        return DateTimeFormatter.ofPattern(pattern).withZone(zone).format(Instant.now());
    }

    @Api("自定义格式化")
    public static String format(String pattern, Instant instant) {
        return DateTimeFormatter.ofPattern(pattern).withZone(ZoneId.systemDefault()).format(instant);
    }

    @Api("自定义格式化")
    public static String format(String pattern, ZoneId zone, Instant instant) {
        return DateTimeFormatter.ofPattern(pattern).withZone(zone).format(instant);
    }

    @Api("自定义格式化")
    public static String format(String pattern, long timeStamp) {
        return DateTimeFormatter.ofPattern(pattern).withZone(ZoneId.systemDefault()).format(Instant.ofEpochMilli(timeStamp));
    }

    @Api("自定义格式化")
    public static String format(String pattern, ZoneId zone, long timeStamp) {
        return DateTimeFormatter.ofPattern(pattern).withZone(zone).format(Instant.ofEpochMilli(timeStamp));
    }


    /**
     * 抽取后的逻辑不易读
     */
    @Api("格式化时间间隔")
    public static String duration(long time) {

        long ms = time;
        long dd = ms / 86400000;
        ms = ms % 86400000;
        long hh = ms / 3600000;
        ms = ms % 3600000;
        long mm = ms / 60000;
        ms = ms % 60000;
        long ss = ms / 1000;
        ms = ms % 1000;
        StringBuilder builder = new StringBuilder();

        boolean contains = false;

        if (dd > 0) {
            contains = true;
            builder.append(dd);
            builder.append(" - ");
        }

        if (hh > 0) {
            contains = true;
            if (builder.length() == 0) {
                builder.append(hh);
            } else {
                builder.append(String.format("%02d", hh));
            }
            builder.append(":");
        } else if (contains) {
            builder.append("00:");
        }

        if (mm > 0) {
            contains = true;
            if (builder.length() == 0) {
                builder.append(mm);
            } else {
                builder.append(String.format("%02d", mm));
            }
            builder.append(":");
        } else if (contains) {
            builder.append("00:");
        }

        if (ss > 0) {
            contains = true;
            if (builder.length() == 0) {
                builder.append(ss);
            } else {
                builder.append(String.format("%02d", ss));
            }
        } else if (contains) {
            builder.append("00");
        }

        if (ms > 0) {
            contains = true;
            builder.append(".");
            if (builder.length() == 0) {
                builder.append(ms);
            } else {
                builder.append(String.format("%03d", ms));
            }
        }

        if (contains) {
            return builder.toString();
        } else {
            return "0.000";
        }
    }


    // ==================================================================================================


    @Api(value = "以系统时间偏移计算是否是今天", attention = "建议只用于lastModify判断")
    public static boolean isToday(long timeStamp) {
        int currentDay = LocalDate.now().getDayOfYear();
        int targetDay = LocalDate.ofInstant(Instant.ofEpochMilli(timeStamp), FurryBlack.SYSTEM_OFFSET).getDayOfYear();
        return currentDay == targetDay;
    }


    @Api("获取休眠到明天的毫秒数")
    public static long nextDayDuration() {
        Instant currentTime = LocalDateTime.now().toInstant(ZoneOffset.UTC);
        Instant targetTime = LocalDate.now().plusDays(1).atStartOfDay().toInstant(ZoneOffset.UTC);
        return targetTime.toEpochMilli() - currentTime.toEpochMilli();
    }


    // ==================================================================================================


}

File: src/main/kotlin/studio/blacktech/furryblack/core/enhance/TimeTool.kt
/*
 * Copyright (C) 2021 Alceatraz @ BlackTechStudio
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms from the BTS Anti-Commercial & GNU Affero General
 * Public License as published by the Free Software Foundation, either
 * version 3 from the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty from
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * BTS Anti-Commercial & GNU Affero General Public License for more details.
 *
 * You should have received a copy from the BTS Anti-Commercial & GNU Affero
 * General Public License along with this program.
 *
 */
@file:Suppress("unused")

package studio.blacktech.furryblack.core.enhance

import studio.blacktech.furryblackplus.FurryBlack
import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.temporal.TemporalAccessor

object TimeTool {

  private val DATETIME = pattern("yyyy-MM-dd HH:mm:ss")
  private val DATE = pattern("yyyy-MM-dd")
  private val TIME = pattern("HH:mm:ss")

  private fun Long.to() = Instant.ofEpochMilli(this)

  private infix fun TemporalAccessor.to(format: String): String = pattern(format).format(this)
  private infix fun TemporalAccessor.to(formatter: DateTimeFormatter): String = formatter.format(this)

  @JvmStatic fun pattern(pattern: String): DateTimeFormatter = pattern(pattern, ZoneId.systemDefault())
  @JvmStatic fun pattern(pattern: String, zone: ZoneId): DateTimeFormatter = DateTimeFormatter.ofPattern(pattern).withZone(zone)

  @JvmStatic fun date() = Instant.now() to DATE
  @JvmStatic fun date(epoch: Long) = epoch.to() to DATE
  @JvmStatic fun date(instant: Instant) = instant to DATE

  @JvmStatic fun time() = Instant.now() to TIME
  @JvmStatic fun time(epoch: Long) = epoch.to() to TIME
  @JvmStatic fun time(instant: Instant) = instant to TIME

  @JvmStatic fun datetime() = Instant.now() to DATETIME
  @JvmStatic fun datetime(epoch: Long) = epoch.to() to DATETIME
  @JvmStatic fun datetime(instant: Instant) = instant to DATETIME

  @JvmStatic fun format(pattern: String) = Instant.now() to pattern
  @JvmStatic fun format(pattern: String, epoch: Long) = epoch.to() to pattern
  @JvmStatic fun format(pattern: String, instant: Instant) = instant to pattern

  @JvmStatic fun format(formatter: DateTimeFormatter) = Instant.now() to formatter
  @JvmStatic fun format(formatter: DateTimeFormatter, epoch: Long) = epoch.to() to formatter
  @JvmStatic fun format(formatter: DateTimeFormatter, instant: Instant) = instant to formatter

  @JvmStatic fun timeToTomorrow() = timeToTomorrow(0, 0, 0, 0)

  @JvmStatic fun timeToTomorrow(hh: Int = 0, mm: Int = 0, ss: Int = 0, sss: Int = 0): Long {
    val now = LocalDateTime.now()
    val next = LocalDateTime.of(now.year, now.monthValue, now.dayOfMonth + 1, hh, mm, ss, sss)
    return next.toEpochSecond(FurryBlack.SYSTEM_OFFSET) * 1000
  }

  @Suppress("DuplicatedCode")
  @JvmStatic fun duration(time: Long): String {

    var ms = time

    val dd = ms / 86400000
    ms %= 86400000
    val hh = ms / 3600000
    ms %= 3600000
    val mm = ms / 60000
    ms %= 60000
    val ss = ms / 1000
    ms %= 1000

    val builder = StringBuilder()

    return if (dd > 0) {
      builder.append(dd)
      builder.append("-")
      builder.append(String.format("%02d", hh))
      builder.append(":")
      builder.append(String.format("%02d", mm))
      builder.append(":")
      builder.append(String.format("%02d", ss))
      builder.append(".")
      builder.append(String.format("%03d", ms))
      builder.toString()
    } else if (hh > 0) {
      builder.append(hh)
      builder.append(":")
      builder.append(String.format("%02d", mm))
      builder.append(":")
      builder.append(String.format("%02d", ss))
      builder.append(".")
      builder.append(String.format("%03d", ms))
      builder.toString()
    } else if (mm > 0) {
      builder.append("00:")
      builder.append(String.format("%02d", mm))
      builder.append(":")
      builder.append(String.format("%02d", ss))
      builder.append(".")
      builder.append(String.format("%03d", ms))
      builder.toString()
    } else if (ss > 0) {
      builder.append("00:00:")
      builder.append(String.format("%02d", ss))
      builder.append(".")
      builder.append(String.format("%03d", ms))
      builder.toString()
    } else if (ms > 0) {
      builder.append("0.")
      builder.append(String.format("%03d", ms))
      builder.toString()
    } else {
      "0.000"
    }
  }
}